src/qibo/ui/bloch.py

```diff
@@ -0,0 +1,407 @@
+from logging import warning
+
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.patches import FancyArrowPatch
+from mpl_toolkits.mplot3d import Axes3D, proj3d
+
+from qibo import hamiltonians
+from qibo.config import raise_error
+from qibo.symbols import X, Y, Z
+
+
+class Arrow3D(FancyArrowPatch):
+    def __init__(self, xs, ys, zs, *args, **kwargs):
+        super().__init__((0, 0), (0, 0), *args, **kwargs)
+        self._verts3d = xs, ys, zs
+
+    def do_3d_projection(self, renderer=None):
+        xs3d, ys3d, zs3d = self._verts3d
+        xs, ys, zs = proj3d.proj_transform(xs3d, ys3d, zs3d, self.axes.M)
+        self.set_positions((xs[0], ys[0]), (xs[1], ys[1]))
+
+        return np.min(zs)
+
+
+class Bloch:
+    def __init__(self, state=None, vector=None, point=None):
+        # Size and style
+        self.figsize = [5, 5]
+        self.fontsize = 18
+        self.arrow_style = "-|>"
+        self.arrow_width = 2.0
+        self.mutation_scale = 20
+        self.linewidth = 0.9
+        self.linecolor = "#383838"
+        self.main_alpha = 0.6
+        self.secondary_alpha = 0.2
+
+        # Figure and axis
+        self.fig = plt.figure(figsize=self.figsize)
+        self.ax = self.fig.add_subplot(111, projection="3d", elev=30, azim=30)
+
+        # Bool variable
+        self._shown = False
+
+        # Data
+        self.points = []
+        self.vectors = []
+
+        # Color data
+        self.color_points = []
+        self.color_vectors = []
```

Comment:
Limit the number of local variables as much as possible.

Especially regarding style, use [Matplotlib stylesheet facilities](https://matplotlib.org/stable/users/explain/customizing.html).
In particular, I'd recommend you to establish a default stylesheet specification as a single top-level dictionary constant. I.e.

```py
STYLE = {
    "figsize": [5, 5]
    ...
}
```

And instead of accessing it individually, just bulk set the stylesheet right before plotting. Possibly using a context, to minimize interaction with global variables. E.g.
```py
with mpl.rc_context(STYLE):
    plt.plot(data)
```
https://matplotlib.org/stable/users/explain/customizing.html#temporary-rc-settings

Reply:
I do not work that much with plots, but the idea is to compose stylesheets as dictionaries (e.g. using `dict.update()`), and only apply them at the very end, and temporary (not to affect other plots).